### ISSUE TITLE
Regex message filter (in-chat voting)

### ISSUE THREAD

**Issue by x9136** [IS2511/ChatIS#22]
Hey!

When streamers play GTA V Chaos Mode, viewers vote (1-8) via chat for options on how to prevent the streamer from passing the game. As a result, the chat turns into a set of numbers, which complicates communication with the streamer and between each other. At the same time, in Chaos Mode, the popularity of the options is displayed in a separate block on the screen, so duplication occurs.

It is easy for viewers to avoid this by creating a rule in one of the extensions to filter messages using Regex: ^[1-8]$. I suggest adding the same feature for the chat overlay to avoid duplication and exclude unnecessary info. It will also improve the quality of the chat on VODs, which can be viewed years later.

**Comment by IS2511** [comment]
Considering it's an overlay that is designed to show message from chat, and not exactly for communicating, I don't see a reason to not show messages that are contextually related to the stream and are significant. Even as an optional feature very little people would use this in the regex form... Maybe with a more user-friendly interface it could work.

> the chat turns into a set of numbers, which complicates communication with the streamer and between each other.

If you are using this overlay as a method of communication - you are using it wrong. It can be used as such, but I don't recommend it. Use Chatterino or a Twitch chat *popout* (to also see votes, donos, etc). Primary uses of this overlay as I see it are: eyeballing stream latency, vaguely judging chat mood, maybe confirming your messages are sending, fallback chat visuals for video editors.

> the popularity of the options is displayed in a separate block on the screen, so duplication occurs.

ChatIS doesn't show the popularity of votes so no duplication can possibly happen. Actually, showing votes would be a cool feature. But I get what you mean. The point is meh anyway, so I can counter with my meh point - the voting overlay doesn't show authorship of votes like ChatIS does.

> exclude unnecessary info

Debatable...

> It will also improve the quality of the chat on VODs

Not much of a use-case IMO. Most video editors use [TwitchDownloader](https://github.com/lay295/TwitchDownloader) (fun fact: I added 7tv emote support to it) to download and render chat history, which is much better than ChatIS for this use-case. And most VODs that are still officially available (so on Twitch) have a chat repeat, which is again way better than ChatIS. If VODs are uploaded to YouTube without chat history - too bad. For small streamers the number spam is not going to be that strong anyway, and for big streamers you can't read the whole chat via ChatIS anyway - it's too fast, you can only judge the general mood.

**In conclusion**: I can probably add a regex filter via a URL parameter, but not a priority (too busy right now). The idea of having a voting UI in case chat spams numbers is interesting, worth considering in the future.

**Comment by x9136** [comment]
As far as I can see, I didn't make myself clear. I did not mean that communication takes place through the chat on the screen. I noted this to show that although the numbers make it difficult to chat, most users filter the chat (via 7TV, FFZ or Chatterino) to hide the numbers and communicate. And in ChatIS, this cannot be done so that the chat on the screen is meaningful.

> ChatIS doesn't show the popularity of votes so no duplication can possibly happen.

The key point is to hide the numbers from the stream, because they duplicate the block that is installed with Chaos Mode (it is also on the stream), and also reduces the quality of the chat on VODs.
<details><summary>Chaos Mode votes</summary><img src="https://github.com/IS2511/ChatIS/assets/60105060/66877278-217d-444b-a577-1b44b8f9d87c"></details>

> the voting overlay doesn't show authorship of votes like ChatIS does.

It makes no difference to Chaos Mode.

> Debatable...

I can easily imagine that people would exclude banwords or spam from regular streams. It doesn't have to be Regex, a list is enough.

> Not much of a use-case IMO.

According to my observations, streamers use ChatIS in order not to lose the chat, because they upload full VODs to other places. Probably, they don't have the desire to constantly make highlights so that Twitch doesn't delete VODs after 2 months. It is much more convenient when the chat is recorded simultaneously with the stream than to add it separately each time. Most likely, it is also more convenient for video editors, because there is less work, but not all streamers like chat on the screen (it takes up space, there is no control over the content, etc). As a result, video editors download the chat history separately.

> Even as an optional feature very little people would use this in the regex form...

It's true, even if we're not talking about Regex. I would not waste time on this feature if it requires more than 1-2 hours of development. Although it's hard for me to imagine that this requires a lot of work if the implementation is similar to "Bot Ignore List".

Anyway, I wanted to highlight such a moment, because it is interesting to watch VODs that have a meaningful chat. In the last couple of months, such a filter would have been useful. Most likely, next time it will be useful only on GTA VI.

**Comment by IS2511** [comment]
> I can easily imagine that people would exclude banwords or spam from regular streams.

They should use the appropriate tools, like AutoMod or a bot, that actually moderate chat - ChatIS supports message deletion and user timeout/ban. ChatIS should just reflect whatever is happening in chat. Viewers can opt-out of seeing removed messages on their end, but they can't opt-in to see messages on ChatIS. Ignoring bots is an opt-in for viewers, but a reasonable compromise for ChatIS as most bot messages don't relate to ongoing stream context.

> it is also more convenient for video editors, because there is less work

Yes, but not by much. Also, I hate to bring this up, but if someone is a professional video editor - they are probably rendering chat separately via TwitchDownloader or similar, using on-screen chat is a fallback IMO.

> I would not waste time on this feature if it requires more than 1-2 hours of development.

I'm a bit of a ~~perfectionist~~ researcher 🫠 sometimes and it can take from 1h to infinity tbh. If I don't forget about this on the weekend I'll probably add this. The tedious part is UI. If you are fine with a URL parameter without docs - it can easily be done. Honestly the whole homepage is a mess and I should fully rewrite it.
